Refactored harvester resource search

reaperrr opened the pull request:
Rebase of #7658, with some minor additions by me.

Fixes or at least massively improves #7092, #6269 and #3089.

Rydra committed:
Improved the performance and intelligence of resource harvesting by
refactoring the Harvesters' pathfinding. Now they in first place assess
which is the closest resource inside their search area and then a path is
calculated

Changed the way harvesters find resources by always trying to find the
closest resource to their refinery.

Changed the strategy of finding to find resources in Annulus.

Mailaender commented on OpenRA.Mods.Common/Activities/DeliverResources.cs:
```diff
@@ -46,8 +58,8 @@ public override Activity Tick(Actor self)
-			if (harv.LinkedProc == null)	// no procs exist; check again in 1s.
-				return Util.SequenceActivities(new Wait(25), this);
+			if (harv.LinkedProc == null)	// no refineries exist; check again in 5s.
+				return Util.SequenceActivities(new Wait(125), this);
```
Bonus points for storing these magic values in HarvesterInfo.

Mailaender commented on OpenRA.Mods.Common/Activities/DeliverResources.cs:
```diff
@@ -17,6 +17,14 @@ namespace OpenRA.Mods.Common.Activities
+		// the constructor. Eventually remove!
+		bool loadedVars = false;
+		
 		const int NextChooseTime = 100;
```
[SA1203] All constants must be placed before all non-constants.

Mailaender commented on OpenRA.Mods.Common/Activities/DeliverResources.cs:
```diff
@@ -17,6 +17,14 @@ namespace OpenRA.Mods.Common.Activities
+		// constructed by reflection and cannot add an Actor variable in
+		// the constructor. Eventually remove!
+		bool loadedVars = false;
+		
```
[SP2000] Invalid spacing at the end of the line.

Mailaender commented on OpenRA.Mods.Common/Activities/FindResources.cs:
```diff
@@ -21,6 +20,18 @@ namespace OpenRA.Mods.Common.Activities
+		Mobile mobile;
+		Harvester harv;
+		MobileInfo mobileInfo;
+		IPathFinder pathFinder;
```
You can put a round of `readonly` before those.

reaperrr replied:
They are assigned later in Tick (see comment on ```loadVars``` below), so that's not possible in this case.

Mailaender commented on OpenRA.Mods.Common/Activities/DeliverResources.cs:
```diff
@@ -46,8 +60,8 @@ public override Activity Tick(Actor self)
 
-			if (harv.LinkedProc == null)	// no procs exist; check again in 1s.
-				return Util.SequenceActivities(new Wait(25), this);
+			if (harv.LinkedProc == null)	// no refineries exist; check again in 5s.
```
The value is not hard-coded anymore so better remove the comment.

penev92 replied:
The comment is good. Just account for the modifiable time.

Mailaender commented on OpenRA.Mods.Common/Activities/FindResources.cs:
```diff
@@ -34,103 +45,118 @@ public FindResources(CPos avoidCell)
+				self.QueueActivity(mobile.MoveTo(moveTo, 1));
+				self.SetTargetLine(Target.FromCell(self.World, moveTo), Color.Gray, false);
+
+				var randFrames = 125 + self.World.SharedRandom.Next(-35, 35);
```
I know this carried over, but now I wonder why this isn't just `self.World.SharedRandom.Next(90, 160);`

pchote commented on OpenRA.Mods.Common/Activities/DeliverResources.cs:
```diff
@@ -18,6 +18,15 @@ namespace OpenRA.Mods.Common.Activities
+		HarvesterInfo harvInfo;
+
+		// TODO: This is a hack because this class is sometimes
+		// constructed by reflection and cannot add an Actor variable in
```
Where is it constructed by reflection?  This doesn't sound right to me.

obrakmann replied:
That comment is about the `InitialActivity` actor init in `FreeActor`

reaperrr replied:
I just didn't look up the reason for this approach in ```FindResources``` and wrongly assumed it would apply here, too.

pchote commented on OpenRA.Mods.Common/Activities/FindResources.cs:
```diff
@@ -21,6 +20,18 @@ namespace OpenRA.Mods.Common.Activities
+		IPathFinder pathFinder;
+
+		// TODO: This is a hack because this class is sometimes
+		// constructed by reflection and cannot add an Actor variable in
```
I would prefer it if we could fix this first in a separate pr.  We can replace the `InitialActivity` hack on `FreeActor` with code in `Harvester` that will queue this activity if it isn't suppressed by an ActorInit or a rules flag (and maybe tie this in to whatever logic makes harvs auto-harvest when built normally?)

pchote commented on OpenRA.Mods.Common/Activities/FindResources.cs:
```diff
@@ -34,103 +45,118 @@ public FindResources(CPos avoidCell)
 					return new DeliverResources();
+
+				harv.UnblockRefinery(self);
+				var moveTo = harv.LastHarvestedCell ?? (self.Location + new CVec(0, 4));
```
This looks like an awfully arbitrary constant.  It would be nice if it were at least randomizing the direction.

obrakmann commented on OpenRA.Mods.Common/Activities/FindResources.cs:
```diff
@@ -34,103 +45,118 @@ public FindResources(CPos avoidCell)
+				var randFrames = 125 + self.World.SharedRandom.Next(-35, 35);
+				if (NextActivity != null)
+					return Util.SequenceActivities(NextActivity, new Wait(randFrames), this);
 				else
```
If I'm not mistaken, the `else` branch will never be taken since you're queuing up a `Move` activity just a few lines above.

obrakmann commented on OpenRA.Mods.Common/Activities/FindResources.cs:
```diff
@@ -34,103 +45,118 @@ public FindResources(CPos avoidCell)
 			{
-				if (!territory.ClaimResource(self, path[0]))
-					return Util.SequenceActivities(new Wait(25), new FindResources());
+				if (!territory.ClaimResource(self, closestHarvestablePosition.Value))
```
Don't you need to add a check whether `closestHarvestablePosition` has a value here? Or make it the `else` branch of the previous `if` block.

obrakmann commented on OpenRA.Mods.Common/Pathfinder/PathFinderCacheDecorator.cs:
```diff
@@ -68,19 +68,7 @@ public List<CPos> FindUnitPathToRange(CPos source, SubCell srcSub, WPos target,
-
-				return pb;
-			}
+				return pathFinder.FindPath(search);
```
Why are you removing all the caching?

obrakmann commented on OpenRA.Mods.Common/Traits/Harvester.cs:
```diff
@@ -19,10 +19,13 @@
 	{
 		public readonly string[] DeliveryBuildings = { };
 
+		[Desc("How long to wait until (re-)checking for a nearby DeliveryBuilding if not yet assigned to one.")]
```
Please mention that the value is in ticks

obrakmann commented on OpenRA.Mods.Common/Traits/Harvester.cs:
```diff
@@ -174,23 +181,18 @@ public void UnblockRefinery(Actor self)
-					var moveTo = harv.LastHarvestedCell ?? (deliveryLoc + new CVec(0, 4));
-					self.QueueActivity(mobile.MoveTo(moveTo, 1));
-					self.SetTargetLine(Target.FromCell(self.World, moveTo), Color.Gray, false);
+					var moveTo = LastHarvestedCell ?? (deliveryLoc + new CVec(0, 4));
```
Same comment as above about the hard-coded offset

obrakmann commented on OpenRA.Mods.Common/Traits/Harvester.cs:
```diff
@@ -301,52 +302,46 @@ public void ResolveOrder(Actor self, Order order)
-					LastOrderLocation = loc;
-				}
+				var notify = self.TraitsImplementing<INotifyHarvesterAction>();
+				var next = new FindResources();
```
Please move this up a bit and use `self.QueueActivity(next)`, like you do in `UnblockRefinery`.

reaperrr commented:
Rebased and updated. Kept each step I took in separate commits for now, will squash them when this gets thumb ups.

reaperrr commented:
One issue I came across myself: It appears that FindResources occasionally runs into an infinite loop within a tick, effectively freezing the game for good. That's a showstopper, we need to find and fix that bug before we can consider merging this.

reaperrr commented:
> [00:37] <+antares79> reaperrr: ClosestHarvestablePos seems to return self.Location during the freeze (even though searchFromLoc is not self.Location)

reaperrr committed:
Fix for potential infinite loop freeze

reaperrr commented:
Updated.

obrakmann commented on OpenRA.Mods.Common/Activities/FindResources.cs:
```diff
@@ -21,171 +20,153 @@ namespace OpenRA.Mods.Common.Activities
+		readonly MobileInfo mobileInfo;
+		ResourceLayer resLayer;
+		ResourceClaimLayer territory;
+		IPathFinder pathFinder;
```
These three can be readonly as well

obrakmann commented on OpenRA.Mods.Common/Activities/HarvestResource.cs:
```diff
@@ -0,0 +1,75 @@
+			// Turn to one of the harvestable facings
+			if (harvInfo.HarvestFacings != 0)
+			{
+				var facing = self.Trait<IFacing>().Facing;
```
I guess this trait lookup could be cached.

obrakmann commented:
I noticed that in d2k, harvesters eventually stop harvesting (at least when carryalls are used)

reaperrr committed:
Make refinery unblock cell customisable

reaperrr commented:
Rebased.

Note: I also squashed to 4 original commits together. Commits 1+2 had issues that commit 3 fixed, and commit 4 was mostly about cosmetic stuff and caching trait lookups/activities. I believe in this particular case keeping them separate had no adavantage anymore and would only have made rebasing even harder.

abcdefg30 commented:
Can't really comment on the code. But ingame was ok. :+1: / :white_check_mark:

pchote commented:
Have the infinite loop and d2k issues been fixed yet?

abcdefg30 commented:
I didn't encounter any infinite loops, but the bug in d2k is not fixed, my bad.

reaperrr commented:
The loop issue should be fixed by the 2nd commit (thanks to @obrakmann), but I haven't figured out what causes the d2k issue yet.

pchote commented on OpenRA.Mods.Common/Activities/FindResources.cs:
```diff
@@ -58,91 +57,111 @@ public override Activity Tick(Actor self)
+
+		bool IsHarvestable(Actor self, CPos pos)
+		{
+			if (pos == self.Location)
```
The harvester actively harvests from the cell that it is in, so returning false unconditionally isn't right.  This was the same kind of situation that caused units to blow up if they were standing too close to a bridge that is destroyed or repaired (checking if they can enter the cell they are already in returns false, so the bridge thinks they fell in the water and kills them).

I wouldn't be surprised if this is causing your d2k carryall issue.

reaperrr commented:
Updated.

I changed the fix.
Rather than always considering the own position non-harvestable (problematic when dropped by carryall), it now first checks if `self.Location` is harvestable, and if it is, skips the path search completely and directly returns `self.Location` as `ClosestHarvestablePos`. I still saw a D2k harvester stop once (enabled debug options indicate that it ran into some kind of infinite loop as well, as the target point kept flickering on its own cell), but it doesn't crash and d2k harvesters appear to get stuck far less. This needs confirmation form other testers, of course.

reaperrr commented:
After some more testing, my fix doesn't really do anything different, so d2k harvesters still get stuck.

What I can say for sure now is that the issue - both for the freeze and the stopping d2k harvesters - is that the `ClosestHarvestablePos` check is looped infinitely in the same tick, so the problem is not in the pathfinder, but rather how and when `FindResources` checks `ClosestHarvestablePos`.

reaperrr committed:
Fix D2k harvesters getting stuck on spice field after drop by carryall

reaperrr commented:
Updated, managed to fix (or at least work around) the d2k issue. Fix is a bit hacky, but in my opinion the performance improvement justifies a little hackyness.

Promoting back to Next Release, would go well together with the AI fixes.

reaperrr commented:
Updated.

obrakmann commented:
Looks fine to me now. :+1:

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/7cb7bcf96078e2ccf07d1125c11f80b11602eaf7%5E...7cb7bcf96078e2ccf07d1125c11f80b11602eaf7)
